File: Data Structures & Algorithms/jump-game/submission-6.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if len(nums)==1:
            return True
        if nums[0]==0:
            return False
        nums.reverse()
        pointer=0
        for i, val in enumerate(nums):
            if i-val<=pointer:
                pointer=i
        if pointer==len(nums)-1:
            return True
        return False
            

File: Data Structures & Algorithms/combination-target-sum/submission-6.java
class Solution {
    List<List<Integer>> answer;
    public void dfs(int[] nums,int target,List<Integer> arr,int pointer){
        if (arr.stream().mapToInt(Integer::intValue).sum()==target){
            List<Integer> clone=new ArrayList<>(arr);
            answer.add(clone);
            return;
        }
        if ((pointer>=nums.length)||(arr.stream().mapToInt(Integer::intValue).sum()>target)){
            return;
        }
        if (pointer<nums.length){
            arr.add(nums[pointer]);
            dfs(nums,target,arr,pointer);
            arr.remove(arr.size()-1);
        }
        dfs(nums,target,arr,pointer+1);


    }
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        answer=new ArrayList<>();
        dfs(nums,target,new ArrayList<>(),0);
        return answer;
    }
}
